### PR TITLE
Issue #11091: Fail test execution when input file hasn't specified violation messages

### LINKED ISSUE #11091: Fail test execution if Input file does nor have message on violation line and Check has not signle message
it become clear at https://github.com/checkstyle/checkstyle/pull/11051#issuecomment-999107833
that we do need to enforce it.

TODO:
if Check has not single message, and Input file contain `//violation` marker in it, we should fail validation and advice user to specify message in addition.

### PR DESCRIPTION
**Pull request by Vyom-Yadav**
Resolves #11091

### CODE REVIEW

**Comment by Vyom-Yadav** [comment]
Ignore the failing checks right now, most of them are failing as most of the input files do not have violation messages specified in the `// violation` comment. @romani most of the input files need to be updated to pass this change, it would be better to split this into multiple PR's and update all input files.

File: src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java [hunk 1]
```diff
@@ -344,21 +372,72 @@ else if (useFilteredViolations) {
...
+    private static long getNumberOfMessages(String className) throws ClassNotFoundException {
+        final Class<?> clazz = Class.forName(className);
+        final String messageInitials = "MSG_";
+        return Arrays.stream(clazz.getDeclaredFields())
+                .filter(field -> {
+                    final int modifiers = field.getModifiers();
+                    final String fieldName = field.getName();
+                    return fieldName.startsWith(messageInitials)
+                            && Modifier.isStatic(modifiers)
+                            && Modifier.isFinal(modifiers);
```
Comment (Vyom-Yadav) [inline src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java:423]: Can be removed if there are no other fields in checks starting with `MSG_`

**Comment by romani** [comment]
Please fix violation only for one Check, create new issue to resolve such problems (it will be good first task for new contributors), all other violation move to suppression with comment "until (link to issue)"

File: src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java (hunk 1 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java:423]: All should be final. Good to keep.

**Comment by romani** [comment]
Do not forget to convert PR to ready to review to attract attention of maintainers

**Comment by Vyom-Yadav** [comment]
> Do not forget to convert PR to ready to review to attract attention of maintainers

@romani It is almost ready, was adding a minor commit of adding violation messages to a check when I noticed a problem. Most of the checks use this kind of notation `'{' should be on a new line`
Now if I try to specify this violation message like-
```
// violation ''{' should be on a new line'
```
It captures the wrong group inside-
image
There are a couple of ways of getting around this, first way is changing
```
.*//\s*violation(?:\W+'(.*)')?$    -->    .*//\s*violation(?:\W+\"(.*)\")?$
```
But this way won't work for
Screenshot from 2021-12-31 12-34-15

So the best way according to me is-
```
.*//\s*violation\s+(?:\"(.*)\")?$

OR

.*//\s*violation\s+(?:'(.*)')?$
```
I didn't find any use case of `\W+`, matching a non-word character isn't that useful, it was meant for-
```
// violation - 'some violation message'
```
but we can simply write it as-
```
// violation 'some violation message'
```
What do you think?

**Comment by Vyom-Yadav** [comment]
@romani ping

**Comment by romani** [comment]
please use `.*//\s*violation\s+(?:'(.*)')?$`

File: src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java [hunk 2]
```diff
@@ -87,6 +91,64 @@
     /** The String "(null)". */
     private static final String NULL_STRING = "(null)";
 
+    /**
+     * Checks in which violation message is not specified in input file and have more than
+     * one violation message key.
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java:96]: please create new issue to resolve  all this suppressions.
and reference this issue here as "until #xxxxxx"

**Review by romani — CHANGES_REQUESTED** [review]
please make one more commit in this PR to show how to fix suppresion line (single Check).
such commit will be guide for new contributors as "do the same".

items:

**Comment by Vyom-Yadav** [comment]
@romani Discovered a problem with-
https://github.com/checkstyle/checkstyle/blob/19693113652274bac0e0d631feb366dcfa51445e/src/test/java/com/puppycrawl/tools/checkstyle/bdd/TestInputViolation.java#L76-L86
It escapes certain characters but not all, a violation message like-
`// violation ''[' should be on the previous line'`
Will throw an exception as `[` is a special regex character. This problem also has a couple of solutions but no solution is pleasing.

### **Sol 1**
Manually escape the violation message i.e. 
`// violation ''\[' should be on the previous line'`

### **Sol 2**
Escape all characters is the violation message-
```
public String toRegex() {                          
    String regex = lineNo + ":(?:\\d+:)?\\s.*";    
    if (message != null) {                         
        String rawMessage = Pattern.quote(message);
        regex += rawMessage + ".*";                
    }                                              
    return regex;                                  
}                                                  
```
But doing this drops the functionality of shortening violation message using `.*`, something like won't work
```
// violation ''[' should .* previous line'
```
as the violation message will be concatenated as `\Q'[' should .* previous line\E`, it will conveniently ignore the `.*` present in violation message.

Escaping some special characters in `toRegex()` is very error-prone and can yield unwanted results.
A makeshift arrangement like not escaping `.*` present between words **can** work but is not recommended as it purely depends upon the nature of the violation message. Please share your opinion.

**Comment by romani** [comment]
> Manually escape the violation message i.e.

yes, message is regexp, so here is side effect.

File: src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java (hunk 2 above)
Comment (Vyom-Yadav) [reply src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java:96]: Done

**Review by romani — APPROVED** [review]
ok to merge.

@Vyom-Yadav , thanks a lot for awesome udpdate

**Pull request merged**